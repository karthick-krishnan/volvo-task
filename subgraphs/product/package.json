{
  "name": "subgraph-product",
  "version": "0.0.1",
  "description": "",
  "scripts": {
    "start": "tsc && node ./dist/src/app.js",
    "build": "tsc",
    "test": "jest --detectOpenHandles --forceExit"
  },
  "dependencies": {
    "@apollo/federation": "^0.33.9",
    "@apollo/subgraph": "^0.1.5",
    "@graphql-tools/graphql-file-loader": "^7.3.3",
    "@graphql-tools/load": "^7.5.1",
    "apollo-graphql": "^0.9.5",
    "apollo-server": "^2.9.16",
    "apollo-server-core": "^3.5.0",
    "apollo-server-express": "^2.9.16",
    "apollo-server-testing": "^2.9.16",
    "bunyan": "^1.8.15",
    "class-validator": "^0.13.1",
    "express": "^4.17.1",
    "express-http-context": "^1.2.4",
    "graphql": "^15.7.1",
    "graphql-import": "^1.0.2",
    "graphql-tools": "^8.2.0",
    "joi": "^17.5.0",
    "lodash": "^4.17.21",
    "mysql2": "^2.3.3",
    "reflect-metadata": "^0.1.13",
    "sequelize": "^6.6.5",
    "sequelize-cli": "^6.2.0",
    "sqlite3": "^5.0.2",
    "tedious": "^14.1.0",
    "type-graphql": "^1.1.1",
    "validator": "^13.7.0",
    "winston": "^3.4.0"
  },
  "keywords": [],
  "author": "",
  "license": "MIT",
  "devDependencies": {
    "@types/bunyan": "^1.8.7",
    "@types/cfenv": "^1.2.1",
    "@types/cookie-parser": "^1.4.2",
    "@types/cors": "^2.8.12",
    "@types/express": "^4.17.13",
    "@types/jest": "27.4.0",
    "@types/lodash": "^4.14.173",
    "@types/redis": "^2.8.32",
    "@types/supertest": "^2.0.11",
    "@types/validator": "^13.6.3",
    "jest": "27.4.0",
    "jest-mock-axios": "^4.4.0",
    "jest-trx-results-processor": "^2.2.0",
    "nodemon": "^2.0.12",
    "sequelize-mock": "^0.10.2",
    "serve": "^12.0.0",
    "supertest": "^6.1.6",
    "ts-jest": "^27.1.3",
    "ts-node": "^10.2.1",
    "typescript": "^4.4.2"
  }
}
