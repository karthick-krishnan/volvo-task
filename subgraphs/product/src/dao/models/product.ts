
'use strict';
import { DataTypes, Model } from "sequelize";
import {sequelize} from "./index"
import dbConstants from '../../utils/db-constants'

class Product extends Model {
  public prodId!: string;
  public name!: string;
  public qty!: string;
  public price!: string;
}

Product.init({
    prodId: {
        primaryKey: true,
        allowNull: false,
        type: DataTypes.UUID,
        defaultValue: DataTypes.UUIDV4
      },
      name: {
        allowNull: true,
        type: DataTypes.UUID
      },
      qty: {
        allowNull: true,
        type: DataTypes.NUMBER
      },
      price: {
        allowNull: true,
        type: DataTypes.NUMBER,
      }
}, {
    sequelize,
    tableName: dbConstants.DB.TABLE_NAMES.PRODUCTS,
    timestamps: false
  });

export { Product };