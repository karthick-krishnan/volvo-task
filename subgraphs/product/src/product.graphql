type Product {
  prodId: ID!
  name: String
  qty: Int
  price: Float
}

extend type Query {
  allProducts: [Product]
  product(prodId: ID!): Product
}

extend type Mutation {
  createProduct(name: String, qty: Int, price: Float): Product
}