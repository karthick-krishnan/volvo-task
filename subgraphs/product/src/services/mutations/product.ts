import ProductOperations from "../../dao/operations/product";
import { inputValidation } from "../validators/index";
import { addProductSchema } from "../validators/product-schema";
import Logger from '../../utils/logger'

export const getMutation = (productOperations: any = ProductOperations) => {
    return { 
        createProduct: async (_ : any,args : any) => {
            try {

              inputValidation(args, addProductSchema);
    
              const product = await productOperations.addProduct({
                    name: args.name,
                    qty: args.qty,
                    price: args.price
                })

                return product

            } catch(ex){
                Logger.error('Exception found in Mutation', JSON.stringify(ex));
                throw ex;
              }
            }
        }
    }


export default getMutation()
