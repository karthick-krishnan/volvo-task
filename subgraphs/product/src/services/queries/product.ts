import ProductOperations from "../../dao/operations/product";
import { inputValidation } from "../validators/index";
import { getProductByIdSchema } from "../validators/product-schema";
import Logger from '../../utils/logger'


export const ProductQueries = (productOperations: any = ProductOperations) => {
  return  { 
      allProducts: async() => {
          return await productOperations.getAllProducts();
      },
      product: async (_ : any,args : any) => {
          try{
              inputValidation(args, getProductByIdSchema);
              const result = await productOperations.getProductById(args.prodId);
              return result
          } catch(ex){
            Logger.error('Exception found in Queries', JSON.stringify(ex));
            throw ex;
          }
      }
  }
}


export default ProductQueries();

