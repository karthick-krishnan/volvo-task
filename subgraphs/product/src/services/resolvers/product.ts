import ProductQueries from '../queries/product' 
import ProductMutations from '../mutations/product'
import { GraphQLResolverMap } from 'apollo-graphql'



export const ProductResolvers : GraphQLResolverMap<any> = { 
    Query : ProductQueries,
    Mutation : ProductMutations
}


 