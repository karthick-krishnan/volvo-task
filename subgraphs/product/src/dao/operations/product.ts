import { Product } from "../models/product";
import Logger from '../../utils/logger'
import {InternalServerError} from '../../utils/error-handler'
import {Op} from "sequelize"



class ProductOperations {

    public async addProduct(product: any) : Promise<Product>{
        try {
            return await Product.create(product);
        } catch (ex) {
            Logger.error('Exception found while adding product!!!', ex);
            throw InternalServerError('Exception found in queries');
        }
    }

    public async getAllProducts() : Promise<Product[]> {
        try {
            return await Product.findAll( {
                where: {
                    qty: {
                        [Op.gte] :  1
                    }
                }
            });
        } catch (ex) {
            Logger.error('Exception found while getting all the product!!!!', ex);
            throw InternalServerError('Exception found in queries');
        }
    }

    public async getProductById(prodId: string) :  Promise<Product|null>  {
        try {
            return await Product.findOne({
                where: {
                    prodId : prodId
            }})
        } catch (ex) {
            Logger.error('Exception found while getting product by id', ex);
            throw InternalServerError('Exception found in queries');
        }
    }
}


export default new ProductOperations();