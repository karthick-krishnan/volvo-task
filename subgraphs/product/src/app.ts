import { ApolloServer, gql } from "apollo-server";
import { buildFederatedSchema } from "@apollo/federation";
import { GraphQLError } from "graphql";
import { ProductResolvers } from "./services/resolvers/product";
const port = process.env.APOLLO_PORT || 4000;
const { readFileSync } = require("fs");
import LoggerPlugin from "./utils/logger-plugin";

// this can also be a glob pattern to match multiple files!

const type = `
type Product {
    prodId: ID!
    name: String
    qty: Int
    price: Float
  }
  
  extend type Query {
    allProducts: [Product]
    product(prodId: ID!): Product
  }
  
  extend type Mutation {
    createProduct(name: String, qty: Int, price: Float): Product
  }`;

export const typeDefs = gql(type);

export const server = new ApolloServer({
  schema: buildFederatedSchema({ typeDefs, resolvers: ProductResolvers }),
  plugins: [LoggerPlugin],
});

server
  .listen({ port: port })
  .then(({ url }) => {
    console.log(`ðŸš€ product subgraph ready at ${url}`);
  })
  .catch((err) => {
    console.error(err);
  });
