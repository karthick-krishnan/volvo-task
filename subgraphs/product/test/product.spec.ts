import { ApolloServer } from 'apollo-server';
import { buildFederatedSchema} from '@apollo/federation';
import { ProductQueries } from '../src/services/queries/product';
import { getMutation } from '../src/services/mutations/product';
import { typeDefs } from '../src/app';
import { mockProductOperations, prodId, createdProdId } from './mock';
import { createTestClient} from 'apollo-server-testing';

let server: ApolloServer;

const resolvers = {
  Query : ProductQueries(mockProductOperations),
  Mutation : getMutation(mockProductOperations)
}

const getAllProductsQuery = `query AllProductsQuery {
  allProducts {
    name
  }
}
`
const getProductsQuery = `query ProductQuery($prodId: ID!) {
  product(prodId: $prodId) {
    prodId
    name
  }
}
`
const createProductMutation = `mutation Mutation($name: String, $price: Float, $qty: Int) {
  createProduct(name: $name, price: $price, qty: $qty) {
    name
    prodId
    price
  }
}
`

const type =`
type Product {
    prodId: ID!
    name: String
    qty: Int
    price: Float
  }
  
  extend type Query {
    allProducts: [Product]
    product(prodId: ID!): Product
  }
  
  extend type Mutation {
    createProduct(name: String, qty: Int, price: Float): Product
  }`

// export const typeDefs = gql(type);


 server = new ApolloServer(
    { schema: buildFederatedSchema({typeDefs, resolvers: resolvers }),     
  
  },
);

console.log('server', server);

beforeAll(() => {


})

afterAll((done) => {
  server.stop()
  done()
})

describe("Products subgraph", () => {
  it("Gets all the products", async () => {
    const res = await server.executeOperation({ query: getAllProductsQuery });
    expect(res).toMatchSnapshot();
  });

  it("Gets products based on the id", async () => {
    const res = await server.executeOperation({ query: getProductsQuery, variables: { prodId } });
    expect(res).toMatchSnapshot();
  });

  it("Matches the requested product id", async () => {
    const res: any = await server.executeOperation({ query: getProductsQuery, variables: { prodId } });
    expect(res.data.product.prodId).toBe(prodId);
  });

  it("Creates new product", async () => {
    const res: any = await server.executeOperation({ query: createProductMutation, variables: { name: "Ipad Pro", qty: 25, price: 999.12 } });
    expect(res).toMatchSnapshot();
  })

  it("Created product details", async () => {
    const res: any = await server.executeOperation({ query: createProductMutation, variables: { name: "Ipad Pro", qty: 25, price: 999.12 } });
    expect(res.data.createProduct.prodId).toBe(createdProdId);
    expect(res.data.createProduct.name).toBe("Ipad Pro");
    expect(res.data.createProduct.price).toBe(999.12);
  })
});
