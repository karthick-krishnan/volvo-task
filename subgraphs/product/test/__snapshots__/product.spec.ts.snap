// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Products subgraph Creates new product 1`] = `
Object {
  "data": Object {
    "createProduct": Object {
      "name": "Ipad Pro",
      "price": 999.12,
      "prodId": "a44b434e-7a16-11ec-90d6-0242ac120003",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Products subgraph Gets all the products 1`] = `
Object {
  "data": Object {
    "allProducts": Array [
      Object {
        "name": "MacBook Pro",
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Products subgraph Gets products based on the id 1`] = `
Object {
  "data": Object {
    "product": Object {
      "name": "MacBook Pro",
      "prodId": "607ac5b8-7a16-11ec-90d6-0242ac120003",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
